internal/sail/client/client.go

```diff
@@ -71,34 +89,44 @@ func (s *SailClient) Connect(ctx context.Context) error {
 
 	connectURL := s.addr
 	connectURL.Path = "/share"
-	conn, _, err := websocket.DefaultDialer.DialContext(ctx, connectURL.String(), header)
+	conn, err := s.dialer.DialContext(ctx, connectURL.String(), header)
 	if err != nil {
 		return err
 	}
 	s.setConnection(ctx, conn)
 	return nil
 }
 
-func (s *SailClient) OnChange(ctx context.Context, st store.RStore) {
-	if s.initDone {
-		return
+func (s *SailClient) init(ctx context.Context) error {
+	if s.addr.Empty() {
+		return nil
 	}
-	defer func() {
+
+	return s.Connect(ctx)
+}
+
+func (s *SailClient) OnChange(ctx context.Context, st store.RStore) {
+	if !s.initDone {
 		s.initDone = true
-	}()
 
-	if s.addr.Empty() {
-		return
+		// TODO(nick): To get an end-to-end connection working, we're just
+		// going to connect to the Sail server on startup. Eventually this
+		// should be changed to connect on user action.
+		err := s.init(ctx)
+		if err != nil {
+			st.Dispatch(store.NewErrorAction(errors.Wrap(err, "SailClient")))
+		}
 	}
 
-	// TODO(nick): To get an end-to-end connection working, we're just
-	// going to connect to the Sail server on startup. Eventually this
-	// should be changed to connect on user action.
-	err := s.Connect(ctx)
-	if err != nil {
-		st.Dispatch(store.NewErrorAction(errors.Wrap(err, "SailClient")))
+	if !s.isConnected() {
```

Comment:
If there was no error but we're also not connected is that an error condition?

Reply:
yes! tho i expect all this code to go away as the code gets fleshed out, see the big TODO a couple lines above this

Reply:
Sounds good to me!